src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs

```diff
@@ -814,8 +814,18 @@ private async Task<WorkUnitResult> ExecuteInstantiatedTask(ITaskExecutionHost ta
                         }
                     }
                 }
-                catch (Exception ex) when (!ExceptionHandling.IsCriticalException(ex) && Environment.GetEnvironmentVariable("MSBUILDDONOTCATCHTASKEXCEPTIONS") != "1")
+                catch (Exception ex)
                 {
+                    if (ExceptionHandling.IsCriticalException(ex) || Environment.GetEnvironmentVariable("MSBUILDDONOTCATCHTASKEXCEPTIONS") == "1")
+                    {
+                        taskLoggingContext.LogFatalTaskError(
+                            ex,
+                            new BuildEventFileInfo(_targetChildInstance.Location),
+                            _taskNode.Name);
+
+                        throw new CriticalTaskException(ex);
```

Comment:
Is there not a potential problem allocating this on the OOM path?

Reply:
Yes - it can lead to crash with uncatchable OOM.
But that is same to the situation where we logged this exception in `RequestBuilder` - as the logging itself here is more costly (needs to fetch the resource string and format the error message). So technically we are not changing the odds of that eventuality much. 

In optimistic (and hopefully typical) scenario the OOM in a custom task is caused by allocations within that task - either by bigger blocks allocations (where the last block will lead to OOM, but will leave smaller chunks still available) or by smaller chunks allocations that will became unrooted after the stack unrolling, GC will free up that memory. So those two allocations (logging and wrapping exceptions) should be accomodable.
We do not have other greedy threads running - so the available memory should not get depleted asynchronously.

cc @janvorli for runtime perspective - in case I'm wrong in my takeaways from our offline discussion

Reply:
To make the odds a bit better, you could preallocate the `BuildEventFileInfo` and `CriticalTaskException` instances. But there are possibilities of other allocations occurring in the runtime under the hoods, possibly even for jitting the code. So in general, after the OOM exception is thrown, all you do is just a best effort.